### ISSUE TITLE
Support updates with aggregation pipelines

### ISSUE THREAD

**Issue by bfreuden** [vert-x3/vertx-mongo-client#236]
#### Describe the feature 

MongoDB supports updates with aggregation pipelines since 4.2.
https://docs.mongodb.com/manual/tutorial/update-documents-with-aggregation-pipeline/

I propose to add support for updates with aggregation pipelines in the Vert.x MongoDB client.

#### Use cases

The typical use case is updating one or many documents using a logic that is based on fields of the document.
For instance: you would to cut 20% of the price of items matching a certain criteria (the logic is based on the existing "price" field that must be updated).

This feature would allow Vert.x applications to update large amount of documents without having to get them out of the database, process them in Java, then inserting them back into the database.

#### Contribution

I am volunteering for implementing this feature but I will definitely need some guidance because this would be my first contribution to Vert.x.

I already have a test running successfully (see code snippet below) but there is a problem though: the flapdoodle embedded MongoDB (version 4.0) does not support updates with aggregation pipelines (introduced in 4.2) so the test is running against an external MongoDB 4.2.10 server.

```java
  @Test
  public void testUpdateMongoClientUpdateResultUpsertIdResponse() throws Exception {
    String collection = "randomCollection";
    mongoClient.insert(collection, new JsonObject().put("price", 10).put("quantity", 1), onSuccess(id -> {
      mongoClient.insert(collection, new JsonObject().put("price", 20).put("quantity", 2), onSuccess(id2 -> {
        mongoClient.insert(collection, new JsonObject().put("price", 30).put("quantity", 10), onSuccess(id3 -> {
          mongoClient.updateCollectionWithOptions(collection,
            // reduce price of low quantity items
            new JsonObject().put("quantity", new JsonObject().put("$lte", 2)),
            new JsonArray().add(new JsonObject().put("$set", new JsonObject().put("price", new JsonObject().put("$subtract", new JsonArray().add("$price").add(2))))),
            new UpdateOptions(),onSuccess(res -> {
              assertEquals(2, res.getDocModified());
              assertEquals(2, res.getDocMatched());
              testComplete();
            }));
        }));
      }));
    }));
    await();
  }
```

**Comment by bfreuden** [comment]
Hi,
If you want to have a look, the first commit is here:
https://github.com/bfreuden/vertx-mongo-client/commit/9dcbded69a1385c0774aceaca7e3e7aa5cffbac4

Flapdoodle still does not support MongoDB 4.2 but there is a beginning of support for MongoDB 4.4.1.
So I upgraded Flapdoodle to 3.0.0, MongoDB to 4.4.1, and hardcoded the download for Ubuntu because of this Flapdoodle bug:
https://github.com/flapdoodle-oss/de.flapdoodle.embed.mongo/issues/327

All tests are passing (including the new one performing an update with an aggregation pipeline).

**Comment by vietj** [comment]
I think this project should be switched to test container instead of flapdoodle

**Comment by vietj** [comment]
@bfreuden do you mind first doing a PR upgrading this project to test containers?

**Comment by bfreuden** [comment]
Hi Julien, 
Thanks for you reply. 
I can do that!
It will require docker on the build machine though. Is it ok?
Should I first create a dedicated github issue like "Replace flapdoodle with  testcontainers"?

**Comment by bfreuden** [comment]
@vietj as I wanted to proceed, I assumed "yes" and "no" :)

**Comment by vietj** [comment]
@bfreuden made a small comment on the testcontainers PR to bump to 1.15.2 and I'll merge it asap, then we can take care of your contribution

**Comment by vietj** [comment]
is there a PR for this issue ?

**Comment by bfreuden** [comment]
@vietj yes I submitted this PR:
https://github.com/vert-x3/vertx-mongo-client/pull/251

**Comment by bfreuden** [comment]
It required to bump MongoDB version to 4.2 (first version having this feature).

**Comment by vietj** [comment]
ok, can you open also a separate issue for mongo version upgrade and reference it ?

**Comment by bfreuden** [comment]
@vietj Sure!
Updates with aggregation pipelines require MongoDB 4.2 in uni test.
This is the purpose of this ticket: #253 
The #251 PR addresses both upgrading MongoDB version to 4.2, and support for updates with aggregation pipelines.

**Comment by bfreuden** [comment]
@viet not sure where we stand regarding this issue. Let me know if you need something else from me.

**Comment by vietj** [comment]
it needs to be reviewed I think @bfreuden

**Comment by vietj** [comment]
where is the PR for the version update ?

**Comment by vietj** [comment]
I can't see a PR for 4.2 upgrade of mongo version, am I missing soemthing ?

**Comment by bfreuden** [comment]
I understand it must be reviewed. No worries and no need to rush.
The #251 PR addresses both upgrading MongoDB version to 4.2, and support for updates with aggregation pipelines.
Do you need a separate PR for the MongoDB version upgrade?

**Comment by tsegismont** [comment]
@bfreuden can we close this one? Otherwise, what is missing after #254 has been merged?

**Comment by bfreuden** [comment]
@tsegismont yes thank you: I don't think there's anything missing.
